src/lib/genesis_ledger_helper/genesis_ledger_helper.ml

```diff
@@ -46,110 +50,420 @@ module Tar = struct
+              match%bind
+                Cache_dir.load_from_s3 [s3_path] [local_path] ~logger
+              with
+              | Ok () ->
+                  file_exists hash_filename Cache_dir.s3_install_path
+              | Error e ->
+                  Logger.info ~module_:__MODULE__ ~location:__LOC__ logger
+                    "Could not download genesis ledger from $uri: $error"
+                    ~metadata:
+                      [ ("uri", `String s3_path)
+                      ; ("error", `String (Error.to_string_hum e)) ] ;
+                  return None ) )
+      | None ->
+          return None
     in
+    match hash_filename with
+    | Some filename ->
+        return (Some filename)
+    | None -> (
+      match config.base with
+      | Hash hash ->
+          assert (Some hash = config.hash) ;
+          return None
+      | Accounts (_accounts, None) ->
+          return None
+      | Accounts (_accounts, Some {name; newer_than}) ->
+          let newer_than =
+            Genesis_constants.validate_time (Some newer_than)
+            |> Result.map_error ~f:(fun str ->
+                   Error.of_string ("Could not parse 'newer_than': " ^ str) )
+            |> Or_error.ok_exn
+          in
+          let named_filename =
+            named_filename ~constraint_constants
+              ~num_accounts:config.num_accounts ("custom_" ^ name)
+          in
+          Deferred.List.find_map search_paths ~f:(fun path ->
+              match%bind file_exists named_filename path with
+              | None ->
+                  return None
+              | Some filename ->
+                  let%map {mtime; _} = Unix.stat filename in
+                  if Time.(mtime >= newer_than) then Some filename
+                  else (
+                    Logger.info ~module_:__MODULE__ ~location:__LOC__ logger
+                      "Not using ledger file $path, because it was modified \
+                       at $modification_time, before $desired_time"
+                      ~metadata:
+                        [ ("path", `String filename)
+                        ; ("modification_time", `String (Time.to_string mtime))
+                        ; ("desired_time", `String (Time.to_string newer_than))
+                        ] ;
+                    None ) )
+      | Named name ->
+          let named_filename =
+            named_filename ~constraint_constants
+              ~num_accounts:config.num_accounts name
+          in
+          Deferred.List.find_map ~f:(file_exists named_filename) search_paths )
+
+  let load_from_tar ~logger
+      ~(constraint_constants : Genesis_constants.Constraint_constants.t)
+      ?accounts filename =
+    Logger.info ~module_:__MODULE__ ~location:__LOC__ logger
+      "Loading genesis ledger from $path"
+      ~metadata:[("path", `String filename)] ;
+    let dirname = Uuid.to_string (Uuid_unix.create ()) in
+    let%bind () = Unix.mkdir ~p:() dirname in
```

Comment:
Create this in the config directory? because we know we have write permissions there